src/app/clusters/commodity-tariff-server/CommodityTariffAttrsDataMgmt.cpp

```diff
@@ -0,0 +1,785 @@
+    return (lhs.Value() == rhs.Value());
+}
+
+/**
+ * @brief The comparator for specific struct types wrapped to Optional<Nullable<Struct>>
+ */
+template <typename T>
+static bool OptionalNullableStructsAreEqual(const Optional<DataModel::Nullable<T>> & lhs,
+                                            const Optional<DataModel::Nullable<T>> & rhs)
+{
+    // Both missing -> true
+    if (!lhs.HasValue() && !rhs.HasValue())
+    {
+        return true;
+    }
+
+    // One present, one missing -> false
+    if (lhs.HasValue() != rhs.HasValue())
+    {
+        return false;
+    }
+
+    // Both present and both null -> true
+    if (lhs.Value().IsNull() && rhs.Value().IsNull())
+    {
+        return true;
+    }
+
+    // One null, one not null -> false
+    if (lhs.Value().IsNull() != rhs.Value().IsNull())
+    {
+        return false;
+    }
+
+    // Both present and not null -> return false if equal
+    return (lhs.Value().Value() == rhs.Value().Value());
+}
+
+}; // namespace CommonUtilities
+
+CHIP_ERROR TariffUnitDataClass::Validate(const ValueType & aValue) const
+{
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR StartDateDataClass::Validate(const ValueType & aValue) const
+{
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DefaultRandomizationOffsetDataClass::Validate(const ValueType & aValue) const
+{
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DefaultRandomizationTypeDataClass::Validate(const ValueType & aValue) const
+{
+    return CHIP_NO_ERROR;
+}
+
+// TariffInfoDataClass
+
+CHIP_ERROR TariffInfoDataClass::Validate(const ValueType & aValue) const
+{
+    if (!aValue.IsNull())
+    {
+        const PayloadType & newValue = aValue.Value();
+
+        if (!newValue.tariffLabel.IsNull() && newValue.tariffLabel.Value().size() > kTariffInfoMaxLabelLength)
+            return CHIP_ERROR_INVALID_ARGUMENT;
+
+        if (!newValue.providerName.IsNull() && newValue.providerName.Value().size() > kTariffInfoMaxProviderLength)
+            return CHIP_ERROR_INVALID_ARGUMENT;
+
+        if (!newValue.blockMode.IsNull())
+        {
+            VerifyOrReturnError(EnsureKnownEnumValue(newValue.blockMode.Value()) != BlockModeEnum::kUnknownEnumValue,
+                                CHIP_ERROR_INVALID_ARGUMENT);
+        }
+
+        if (CommonUtilities::HasFeatureInCtx(static_cast<TariffUpdateCtx *>(mAuxData), CommodityTariff::Feature::kPricing))
+        {
+            if (newValue.currency.HasValue() && !newValue.currency.Value().IsNull())
+            {
+                const auto & currency = newValue.currency.Value().Value();
+                if (currency.currency >= kMaxCurrencyValue)
+                    return CHIP_ERROR_INVALID_ARGUMENT;
+            }
+            else
+            {
+                // If the pricing feature is enabled, the currency field is required!
+                return CHIP_ERROR_INVALID_ARGUMENT;
+            }
+        }
+    }
+    else
+    {
+        // The tariff info is required, but not present.
```

Comment:
I don't get it.  TariffInfo is allowed to be null.

At this point either you are not implementing the spec I am reading, or something else is seriously wrong here, but either way I am stopping here (after reading through the header for AttrsDataMgmt but nothing else so far).  I fully expect that actually writing some tests that exercise this code will turn up a bunch of changes that need to happen.